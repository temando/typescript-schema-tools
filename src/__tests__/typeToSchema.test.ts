import { mkdir, remove } from 'fs-extra';
import { join } from 'path';
import { ITjsSchema, saveExports, saveSchemas, typesToSchemas } from '..';

/** Whether to delete created files, for testing */
const DELETE_TEMP = true;

describe('typeToSchema', () => {
  const typeFixturesFilePath = join(__dirname, './fixtures/types.ts');
  const tempDirectory = join(__dirname, './__temp__');

  async function ensureTemp () { try { await mkdir(tempDirectory); } catch { /**/ } }
  async function removeTemp () { if (DELETE_TEMP) { await remove(tempDirectory); } }

  beforeAll(ensureTemp);
  afterAll(removeTemp);

  describe('typesToSchemas()', () => {
    it('produces a single schema from one file', async () => {
      const { errors, schemas: [schema] } = await typesToSchemas({
        fromFiles: [typeFixturesFilePath],
        types: { Test: 'ITest' },
      });

      expect(errors).toBeFalsy();
      expect(schema).toMatchSnapshot('singleSchema');
    });

    it('produces multiple schemas from one file', async () => {
      const { errors, schemas } = await typesToSchemas({
        fromFiles: [join(__dirname, './fixtures/types.ts')],
        types: {
          Test: 'ITest',
          Test2: 'ITest2',
        },
      });

      expect(errors).toBeFalsy();
      expect(schemas).toMatchSnapshot('multipleSchemas');
    });
  });

  describe('saveSchemas()', () => {
    let singleSchema: ITjsSchema;
    let manySchemas: ITjsSchema[];

    beforeAll(async () => {
      const { schemas: [schema] } = await typesToSchemas({
        fromFiles: [typeFixturesFilePath],
        types: { Test: 'ITest' },
      });

      singleSchema = schema;

      const { schemas } = await typesToSchemas({
        fromFiles: [typeFixturesFilePath],
        types: { Test: 'ITest', Test2: 'ITest2' },
      });

      manySchemas = schemas;

    });

    it('saves a single schema as default export to a .ts file', async () => {
      const fileName = `singleSchema`;
      const fullFilePath = join(tempDirectory, `${fileName}.ts`);

      await saveSchemas({
        asDefaultExport: true,
        directory: tempDirectory,
        name: fileName,
        format: 'ts',
        schemas: [singleSchema],
      });

      const exported = require(fullFilePath);

      expect(exported).toMatchSnapshot('singleSchemaDefaultExport');
    });

    it('saves a single schema to a .json file', async () => {
      const fileName = `singleSchemaJson`;
      const fullFilePath = join(tempDirectory, `${fileName}.json`);

      await saveSchemas({
        asDefaultExport: true,
        directory: tempDirectory,
        name: fileName,
        format: 'json',
        schemas: [singleSchema],
      });

      const exported = require(fullFilePath);

      expect(exported).toMatchSnapshot('singleSchemaJson');
    });

    it('correctly fails to save multiple schema as a default export', async () => {
      let error: Error | undefined;

      await saveSchemas({
        asDefaultExport: true,
        directory: tempDirectory,
        name: 'thisShouldNotBeSaved',
        format: 'ts',
        schemas: manySchemas,
      }).catch((err) => error = err);

      expect(error).toBeInstanceOf(Error);
    });

    it('saves multiple schema to a .ts file', async () => {
      const fileName = `manySchemas`;
      const fullFilePath = join(tempDirectory, `${fileName}.ts`);

      await saveSchemas({
        asDefaultExport: false,
        directory: tempDirectory,
        name: fileName,
        format: 'ts',
        schemas: manySchemas,
      });

      const exported = require(fullFilePath);

      expect(exported).toMatchSnapshot('manySchemas');
    });
  });

  describe('saveExports()', () => {
    it('creates a file with imports and exports', async () => {
      const fileName = `exports`;
      const fullFilePath = join(tempDirectory, fileName);

      /** This is saving to ./__temp__/exports.ts */
      await saveExports({
        name: fileName,
        directory: tempDirectory,
        getImportPath: (name) => `../fixtures/imports/${name}`,
        exports: ['a', 'b'],
      });

      const exports = require(fullFilePath);

      expect(exports).toMatchSnapshot('saveExports');
    });
  });

});
