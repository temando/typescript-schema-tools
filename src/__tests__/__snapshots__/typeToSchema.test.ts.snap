// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`typeToSchema TypeSchemaBuilder compiles and saves multiple combined schemas: builderTogetherOne 1`] = `
Object {
  "builderTogetherOne": Object {
    "$schema": "http://json-schema.org/draft-06/schema#",
    "additionalProperties": false,
    "properties": Object {
      "a": Object {
        "description": "a",
        "minimum": 1,
        "type": "number",
      },
    },
    "required": Array [
      "a",
    ],
    "type": "object",
  },
  "builderTogetherTwo": Object {
    "$schema": "http://json-schema.org/draft-06/schema#",
    "additionalProperties": false,
    "definitions": Object {
      "IRefMe": Object {
        "additionalProperties": false,
        "properties": Object {
          "a": Object {
            "additionalProperties": false,
            "properties": Object {
              "c": Object {
                "enum": Array [
                  1,
                ],
                "type": "number",
              },
            },
            "required": Array [
              "c",
            ],
            "type": "object",
          },
        },
        "required": Array [
          "a",
        ],
        "type": "object",
      },
    },
    "properties": Object {
      "c": Object {
        "enum": Array [
          1,
          2,
          3,
        ],
        "type": "number",
      },
      "reffed": Object {
        "items": Object {
          "$ref": "#/definitions/IRefMe",
        },
        "type": "array",
      },
    },
    "required": Array [
      "c",
      "reffed",
    ],
    "type": "object",
  },
}
`;

exports[`typeToSchema TypeSchemaBuilder compiles and saves multiple combined schemas: builderTogetherTwo 1`] = `
Object {
  "builderTogetherOne": Object {
    "$schema": "http://json-schema.org/draft-06/schema#",
    "additionalProperties": false,
    "properties": Object {
      "a": Object {
        "description": "a",
        "minimum": 1,
        "type": "number",
      },
    },
    "required": Array [
      "a",
    ],
    "type": "object",
  },
  "builderTogetherTwo": Object {
    "$schema": "http://json-schema.org/draft-06/schema#",
    "additionalProperties": false,
    "definitions": Object {
      "IRefMe": Object {
        "additionalProperties": false,
        "properties": Object {
          "a": Object {
            "additionalProperties": false,
            "properties": Object {
              "c": Object {
                "enum": Array [
                  1,
                ],
                "type": "number",
              },
            },
            "required": Array [
              "c",
            ],
            "type": "object",
          },
        },
        "required": Array [
          "a",
        ],
        "type": "object",
      },
    },
    "properties": Object {
      "c": Object {
        "enum": Array [
          1,
          2,
          3,
        ],
        "type": "number",
      },
      "reffed": Object {
        "items": Object {
          "$ref": "#/definitions/IRefMe",
        },
        "type": "array",
      },
    },
    "required": Array [
      "c",
      "reffed",
    ],
    "type": "object",
  },
}
`;

exports[`typeToSchema TypeSchemaBuilder compiles and saves multiple seperate schemas: builderSeperateOne 1`] = `
Object {
  "default": Object {
    "$schema": "http://json-schema.org/draft-06/schema#",
    "additionalProperties": false,
    "properties": Object {
      "a": Object {
        "description": "a",
        "minimum": 1,
        "type": "number",
      },
    },
    "required": Array [
      "a",
    ],
    "type": "object",
  },
}
`;

exports[`typeToSchema TypeSchemaBuilder compiles and saves multiple seperate schemas: builderSeperateTwo 1`] = `
Object {
  "default": Object {
    "$schema": "http://json-schema.org/draft-06/schema#",
    "additionalProperties": false,
    "definitions": Object {
      "IRefMe": Object {
        "additionalProperties": false,
        "properties": Object {
          "a": Object {
            "additionalProperties": false,
            "properties": Object {
              "c": Object {
                "enum": Array [
                  1,
                ],
                "type": "number",
              },
            },
            "required": Array [
              "c",
            ],
            "type": "object",
          },
        },
        "required": Array [
          "a",
        ],
        "type": "object",
      },
    },
    "properties": Object {
      "c": Object {
        "enum": Array [
          1,
          2,
          3,
        ],
        "type": "number",
      },
      "reffed": Object {
        "items": Object {
          "$ref": "#/definitions/IRefMe",
        },
        "type": "array",
      },
    },
    "required": Array [
      "c",
      "reffed",
    ],
    "type": "object",
  },
}
`;

exports[`typeToSchema saveExports() creates a file with imports and exports: saveExports 1`] = `
Object {
  "a": Object {
    "foo": 1,
  },
  "b": Object {
    "bar": 2,
  },
}
`;

exports[`typeToSchema saveSchemas() saves a single schema as default export to a .ts file: singleSchemaDefaultExport 1`] = `
Object {
  "default": Object {
    "$schema": "http://json-schema.org/draft-06/schema#",
    "additionalProperties": false,
    "properties": Object {
      "a": Object {
        "description": "a",
        "minimum": 1,
        "type": "number",
      },
    },
    "required": Array [
      "a",
    ],
    "type": "object",
  },
}
`;

exports[`typeToSchema saveSchemas() saves a single schema to a .json file: singleSchemaJson 1`] = `
Object {
  "name": "Test",
  "schema": Object {
    "$schema": "http://json-schema.org/draft-06/schema#",
    "additionalProperties": false,
    "properties": Object {
      "a": Object {
        "description": "a",
        "minimum": 1,
        "type": "number",
      },
    },
    "required": Array [
      "a",
    ],
    "type": "object",
  },
  "type": "ITest",
}
`;

exports[`typeToSchema saveSchemas() saves multiple schema to a .ts file: manySchemas 1`] = `
Object {
  "Test": Object {
    "$schema": "http://json-schema.org/draft-06/schema#",
    "additionalProperties": false,
    "properties": Object {
      "a": Object {
        "description": "a",
        "minimum": 1,
        "type": "number",
      },
    },
    "required": Array [
      "a",
    ],
    "type": "object",
  },
  "Test2": Object {
    "$schema": "http://json-schema.org/draft-06/schema#",
    "additionalProperties": false,
    "definitions": Object {
      "IRefMe": Object {
        "additionalProperties": false,
        "properties": Object {
          "a": Object {
            "additionalProperties": false,
            "properties": Object {
              "c": Object {
                "enum": Array [
                  1,
                ],
                "type": "number",
              },
            },
            "required": Array [
              "c",
            ],
            "type": "object",
          },
        },
        "required": Array [
          "a",
        ],
        "type": "object",
      },
    },
    "properties": Object {
      "c": Object {
        "enum": Array [
          1,
          2,
          3,
        ],
        "type": "number",
      },
      "reffed": Object {
        "items": Object {
          "$ref": "#/definitions/IRefMe",
        },
        "type": "array",
      },
    },
    "required": Array [
      "c",
      "reffed",
    ],
    "type": "object",
  },
}
`;

exports[`typeToSchema typesToSchemas() produces a schema with an annotated $ref: annotatedReffedSchema 1`] = `
Array [
  Object {
    "name": undefined,
    "schema": Object {
      "$schema": "http://json-schema.org/draft-06/schema#",
      "additionalProperties": false,
      "properties": Object {
        "a": Object {
          "$ref": "test-ref#",
        },
      },
      "required": Array [
        "a",
      ],
      "type": "object",
    },
    "type": "IRefTest",
  },
]
`;

exports[`typeToSchema typesToSchemas() produces a schema with an overridden $ref: reffedSchema 1`] = `
Array [
  Object {
    "name": "Test2",
    "schema": Object {
      "$schema": "http://json-schema.org/draft-06/schema#",
      "additionalProperties": false,
      "definitions": Object {
        "IRefMe": Object {
          "$ref": "ref-me#",
        },
      },
      "properties": Object {
        "c": Object {
          "enum": Array [
            1,
            2,
            3,
          ],
          "type": "number",
        },
        "reffed": Object {
          "items": Object {
            "$ref": "#/definitions/IRefMe",
          },
          "type": "array",
        },
      },
      "required": Array [
        "c",
        "reffed",
      ],
      "type": "object",
    },
    "type": "ITest2",
  },
]
`;

exports[`typeToSchema typesToSchemas() produces a single schema from one file: singleSchema 1`] = `
Object {
  "name": "Test",
  "schema": Object {
    "$schema": "http://json-schema.org/draft-06/schema#",
    "additionalProperties": false,
    "properties": Object {
      "a": Object {
        "description": "a",
        "minimum": 1,
        "type": "number",
      },
    },
    "required": Array [
      "a",
    ],
    "type": "object",
  },
  "type": "ITest",
}
`;

exports[`typeToSchema typesToSchemas() produces multiple schemas from one file: multipleSchemas 1`] = `
Array [
  Object {
    "name": "Test",
    "schema": Object {
      "$schema": "http://json-schema.org/draft-06/schema#",
      "additionalProperties": false,
      "properties": Object {
        "a": Object {
          "description": "a",
          "minimum": 1,
          "type": "number",
        },
      },
      "required": Array [
        "a",
      ],
      "type": "object",
    },
    "type": "ITest",
  },
  Object {
    "name": "Test2",
    "schema": Object {
      "$schema": "http://json-schema.org/draft-06/schema#",
      "additionalProperties": false,
      "definitions": Object {
        "IRefMe": Object {
          "additionalProperties": false,
          "properties": Object {
            "a": Object {
              "additionalProperties": false,
              "properties": Object {
                "c": Object {
                  "enum": Array [
                    1,
                  ],
                  "type": "number",
                },
              },
              "required": Array [
                "c",
              ],
              "type": "object",
            },
          },
          "required": Array [
            "a",
          ],
          "type": "object",
        },
      },
      "properties": Object {
        "c": Object {
          "enum": Array [
            1,
            2,
            3,
          ],
          "type": "number",
        },
        "reffed": Object {
          "items": Object {
            "$ref": "#/definitions/IRefMe",
          },
          "type": "array",
        },
      },
      "required": Array [
        "c",
        "reffed",
      ],
      "type": "object",
    },
    "type": "ITest2",
  },
]
`;
